### PR TITLE
Add auto-updating docs bot

### LINKED ISSUE #314: [Proposal] Automatically regenerate the docs on push to main
### Proposal
Automatically regenerate the docs on push to main.

### Motivation
The committed `.rst` files are sometimes out of date if not explicitly regenerated in a PR, as running `sphinx-apidoc` locally showed differences [on a recent PR](https://github.com/neelnanda-io/TransformerLens/pull/312).

There are inconsistencies in the documentation that is being served through GH Docs and the actual state of the repository. Recently the "torchtyping helper module" was in the documentation, but the module had been removed from the repository. It is proposed to add automatic rst generation in the CI pipeline to keep the documentation up-to-date.

### Pitch
The proposed solution is to add automatic `rst` generation in the CI pipeline. 
This can be accomplished by adding a step in the GH action that runs `sphinx-apidoc` whenever changes are pushed to the repository (I think! CC @dkamm @Aprillion ) 
This will ensure that the documentation stays consistent with the state of the repository.

### Checklist
[x] I have checked that there is no similar [issue](https://github.com/neelnanda-io/transformerlens/issues) in the repo (required)

### PR DESCRIPTION
**Pull request by Smaug123**
# Description

Add a GitHub action which automatically updates the docs on any PR when needed, if permitted by GitHub's rather restrictive permissions model.

Fixes #314.

I have not tested this locally, because… it's not clear how. But see https://github.com/Smaug123/actions-test/pull/1 for an example, where:

* https://github.com/Smaug123/actions-test/pull/1/commits/a3c97534bd05f4596656fc58eb67cdba3611a310 was generated by the pipeline based on https://github.com/Smaug123/actions-test/blob/db69df2ad05602d79bb0242d0c86edd7ac75d170/.github/workflows/update-doc.yml
* there was not a storm of the action repeatedly triggering in response to its own commit
* it triggered once again on a subsequent commit https://github.com/Smaug123/actions-test/pull/1/commits/7d1bde2491ab3dc2bb0dc2e19f591f3b3dfa86ee
* it could be triggered repeatedly as the user kept making changes.

Note that the docs on `main` seem to be out of date, so the check is failing on this PR right now. Raised https://github.com/neelnanda-io/TransformerLens/pull/339 to bring them back up to date.

## Type of change

Invisible to consumers.

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by Smaug123** [comment]
Oh my - I can run new actions on this repo! That makes it considerably easier, although it seems a bit dubious from a security point of view.

**Comment by Smaug123** [comment]
Hmm - looks like this might not cope correctly with forks.

**Comment by Smaug123** [comment]
Sigh, this fails because it needs permission to push to my fork.

**Comment by Smaug123** [comment]
If this were my own project, I'd be happy enough with this as of 17908cb. Have at it, maintainers!

File: .github/workflows/update-doc.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+name: Regenerate docs
+
+on: pull_request
```
Comment (Smaug123) [inline .github/workflows/update-doc.yml:3]: One could consider doing this `on: push` instead.

**Comment by jbloomAus** [comment]
hey Patrick, 

This seems like it might be useful but also like it might be unnecessary/create work. So I'm going to say that we won't accept this for the time being, but I'll try to remember it if people are asking for this in the future. 

I'm keen to accept PR's that increase stability, or solve problems users are having, but I'm wary of premature optimisation as I'm pretty time short and want to preserve attention on more essential changes/issues. 

I'd be keen to accept PRs you make along these lines though!

**Pull request closed without merging**